# Dataset Extraction into Three Excel Files

## 1. Brain Mass & Body Length Data
This file contains measurements of brain mass and body length for 339 individuals across 23 species.

## 2. Head Shape Landmark Data
This file includes head shape data for 23 species using 12 homologous landmarks, formatted in NTS.

## 3. Molecular-Morphological Phylogeny
This file presents a Nexus-format phylogenetic description of the species based on molecular and morphological data.

In [4]:
pip install openpyxl

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [30]:
import pandas as pd
import pandas as pd
from sklearn.preprocessing import LabelEncoder

In [6]:
# Re-read the dataset
with open("Data Pre-Processing\Syngnathid brain size.txt", "r", encoding="utf-8") as file:
    data = file.readlines()

In [7]:
# Initialize lists for each section
brain_mass_data = []
head_shape_data = []
phylogeny_data = []

current_section = None

In [8]:
for line in data:
    line = line.strip()  # Remove extra spaces and newlines

    if "#1: Species Data" in line:
        current_section = "brain_mass"
        continue
    elif "#2 Head morphology data" in line:
        current_section = "head_shape"
        continue
    elif "#3 Molecular-morphological phylogeny" in line:
        current_section = "phylogeny"
        continue

    if current_section == "brain_mass" and line:
        brain_mass_data.append(line.split("\t"))  # Tab-separated values
    elif current_section == "head_shape" and line:
        head_shape_data.append(line.split())  # Space-separated values
    elif current_section == "phylogeny" and line:
        phylogeny_data.append([line])  # Store as text lines

In [9]:
# Convert to DataFrames, ensuring proper structure
brain_mass_columns = ["Species", "Sex", "Length (mm)", "Brain weight (mg)", "Preservative", "Sampled location", "Sample code"]
brain_mass_df = pd.DataFrame(brain_mass_data[1:], columns=brain_mass_columns)

In [10]:
# Head shape data (Assuming the first row contains species names)
head_shape_df = pd.DataFrame(head_shape_data)

# Phylogeny data as text
phylogeny_df = pd.DataFrame(phylogeny_data, columns=["Phylogeny"])

In [11]:
# Save to Excel files
brain_mass_df.to_excel("Data Pre-Processing/brain_mass_body_length.xlsx", index=False)
head_shape_df.to_excel("Data Pre-Processing/head_shape.xlsx", index=False)
phylogeny_df.to_excel("Data Pre-Processing/molecular_phylogeny.xlsx", index=False)

# Data Visualisation of brain_mass_body_length.xlsx

In [12]:
# Load the brain mass and body length dataset
brain_mass_path = "Data Pre-Processing/brain_mass_body_length.xlsx"
brain_mass_df = pd.read_excel(brain_mass_path)

In [49]:
print(len(brain_mass_df))

339


In [13]:
brain_mass_df.head(5)
brain_mass_df.head(-5)

Species Sex  Length (mm)  Brain weight (mg)   
0    Corythoichthys haematopterus   f        112.6               3.27  \
1    Corythoichthys haematopterus   f        115.0               3.97   
2    Corythoichthys haematopterus   f        121.4               4.83   
3    Corythoichthys haematopterus   f        122.4               4.09   
4    Corythoichthys haematopterus   f        128.0               4.29   
..                            ...  ..          ...                ...   
329             Syngnathus typhle   m        204.0              11.30   
330             Syngnathus typhle   m        176.0              10.20   
331             Syngnathus typhle   m        166.0               9.40   
332             Syngnathus typhle   m        208.0              10.60   
333         Trachyhampus serratus   f        205.0               4.70   

    Preservative Sampled location   Sample code  
0       Formalin            Japan           NaN  
1       Formalin            Japan           NaN  
2       Formalin            Japan           NaN  
3       Formalin            Japan           NaN  
4       Formalin            Japan           NaN  
..           ...              ...           ...  
329     Formalin           Sweden           NaN  
330     Formalin           Sweden           NaN  
331     Formalin           Sweden           NaN  
332     Formalin           Sweden           NaN  
333      Ethanol           Taiwan  NMST P110970  

[334 rows x 7 columns]

Will rename 'Sample location' to 'location', 'Brain weight (mg)' to 'Brain_weight', 'Length (mm)' to 'Length'

In [ ]:
# Rename multiple columns for consistency
brain_mass_df.rename(columns={
    "Sample location": "Location",
    "Brain weight (mg)": "Brain_weight",
    "Length (mm)": "Length"
}, inplace=True)

brain_mass_df.to_excel("data/brain_mass_body_length.xlsx", index=False)

In [15]:
#To check if changes made
brain_mass_df.head(5)

Species Sex  Length (mm)  Brain weight (mg)   
0  Corythoichthys haematopterus   f        112.6               3.27  \
1  Corythoichthys haematopterus   f        115.0               3.97   
2  Corythoichthys haematopterus   f        121.4               4.83   
3  Corythoichthys haematopterus   f        122.4               4.09   
4  Corythoichthys haematopterus   f        128.0               4.29   

  Preservative Location Sample code  
0     Formalin    Japan         NaN  
1     Formalin    Japan         NaN  
2     Formalin    Japan         NaN  
3     Formalin    Japan         NaN  
4     Formalin    Japan         NaN

In [16]:
brain_mass_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 339 entries, 0 to 338
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Species            339 non-null    object 
 1   Sex                339 non-null    object 
 2   Length (mm)        339 non-null    float64
 3   Brain weight (mg)  339 non-null    float64
 4   Preservative       339 non-null    object 
 5   Location           339 non-null    object 
 6   Sample code        33 non-null     object 
dtypes: float64(2), object(5)
memory usage: 18.7+ KB


# Inference:
There are 339 values in each variables except 'Code' which only has 33 values.

We will drop 'Sample code' column as it is unnecessary and only has 33 values. Then, we will see the unique values of each column and check if there are any irregularities.

In [17]:
# Drop the "Code" column from the dataset as it is unecessary and save it.
brain_mass_df.drop(columns=["Sample code"], inplace=True, errors="ignore")
brain_mass_df.to_excel("data/brain_mass_body_length.xlsx", index=False)

In [18]:
#To check if changes made
brain_mass_df.head(5)

Species Sex  Length (mm)  Brain weight (mg)   
0  Corythoichthys haematopterus   f        112.6               3.27  \
1  Corythoichthys haematopterus   f        115.0               3.97   
2  Corythoichthys haematopterus   f        121.4               4.83   
3  Corythoichthys haematopterus   f        122.4               4.09   
4  Corythoichthys haematopterus   f        128.0               4.29   

  Preservative Location  
0     Formalin    Japan  
1     Formalin    Japan  
2     Formalin    Japan  
3     Formalin    Japan  
4     Formalin    Japan

In [19]:
brain_mass_df.nunique()

Species               24
Sex                    2
Length (mm)          224
Brain weight (mg)    267
Preservative           2
Location              17
dtype: int64

Now we will see all the unique values in Species, Preservatives and Location

In [20]:
brain_mass_df["Species"].unique()

array(['Corythoichthys haematopterus', 'Corythoichthys intestinalis',
       'Doriyichthys boaja', 'Doriyichthys martensi',
       'Doryhampus dactyliophorus', 'Doryrhamphus japonicus',
       'Entelurus aequoreus', 'Hippichthys penicillus',
       'Hippocampus abdominalis', 'Hippocampus addominalis',
       'Hippocampus comes', 'Hippocampus hippocampus', 'Hippocampus kuda',
       'Hippocampus spinosissimus', 'Hippocampus trimaculatus',
       'Microphis brachyurus', 'Nerophis lumbriciformis',
       'Nerophis ophidion', 'Syngnathoides biaculeatus',
       'Syngnathus abaster', 'Syngnathus acus', 'Syngnathus schlegeli',
       'Syngnathus typhle', 'Trachyhampus serratus'], dtype=object)

Here there is an spelling error in one of the species. The error is "Hippocampus addominalis", the correct one is "Hippocampus abdominalis". We will check the rows which have the incorrect species and then change it.

In [21]:
# Find rows where the incorrect species name "Hippocampus addominalis" appears
incorrect_rows = brain_mass_df[brain_mass_df["Species"] == "Hippocampus addominalis"]
incorrect_rows

Species Sex  Length (mm)  Brain weight (mg) Preservative   
124  Hippocampus addominalis   f        184.5               15.8     Formalin  \
125  Hippocampus addominalis   f        146.7               10.1     Formalin   
126  Hippocampus addominalis   f        156.6               20.2     Formalin   
127  Hippocampus addominalis   m        151.0               10.9     Formalin   
128  Hippocampus addominalis   m        165.0               27.1     Formalin   
129  Hippocampus addominalis   m        136.8               12.1     Formalin   
130  Hippocampus addominalis   m        156.1               14.6     Formalin   

      Location  
124  Australia  
125  Australia  
126  Australia  
127  Australia  
128  Australia  
129  Australia  
130  Australia

In [22]:
# Replace the incorrect name with the correct one "Hippocampus abdominalis" and save it
brain_mass_df["Species"] = brain_mass_df["Species"].replace("Hippocampus addominalis", "Hippocampus abdominalis")
brain_mass_df.to_excel("data/brain_mass_body_length.xlsx", index=False)

To check all the unique values in our data and check if something need to be changed or deleted

In [23]:
length = len(brain_mass_df["Species"].unique())
print("Length of 'Species':", length)
brain_mass_df["Species"].unique()

Length of 'Species': 23


array(['Corythoichthys haematopterus', 'Corythoichthys intestinalis',
       'Doriyichthys boaja', 'Doriyichthys martensi',
       'Doryhampus dactyliophorus', 'Doryrhamphus japonicus',
       'Entelurus aequoreus', 'Hippichthys penicillus',
       'Hippocampus abdominalis', 'Hippocampus comes',
       'Hippocampus hippocampus', 'Hippocampus kuda',
       'Hippocampus spinosissimus', 'Hippocampus trimaculatus',
       'Microphis brachyurus', 'Nerophis lumbriciformis',
       'Nerophis ophidion', 'Syngnathoides biaculeatus',
       'Syngnathus abaster', 'Syngnathus acus', 'Syngnathus schlegeli',
       'Syngnathus typhle', 'Trachyhampus serratus'], dtype=object)

In [24]:
brain_mass_df["Preservative"].unique()

array(['Formalin', 'Ethanol'], dtype=object)

In [25]:
brain_mass_df["Location"].unique()

array(['Japan', 'Malaysia', 'Micronesia', 'Marshall', 'Tokushima (Japan)',
       'Hahajima (Japan)', 'Ioujima (Japan)', 'Izu (Japan)',
       'Miyakejima (Japan)', 'Sweden', 'Australia', 'Okinawa (Japan)',
       'Portugal', 'Otsuchi bay (Japan)', 'Taiwan', 'Surugawan (Japan)',
       'Hyogo (Japan)'], dtype=object)

In [26]:
brain_mass_df.describe()

Length (mm)  Brain weight (mg)
count   339.000000         339.000000
mean    175.572566           8.615929
std      65.259763           5.717701
min      51.500000           1.700000
25%     124.650000           4.505000
50%     162.000000           7.990000
75%     218.500000          10.315000
max     415.000000          32.100000

# Inference

## 1. Species  
There are **23** species in the dataset. They are:  

### **Seahorses (Genus: Hippocampus)**
- *Hippocampus abdominalis*  
- *Hippocampus comes*  
- *Hippocampus hippocampus*  
- *Hippocampus kuda*  
- *Hippocampus spinosissimus*  
- *Hippocampus trimaculatus*  

### **Pipefishes (Various Genera)**
- **Tropical Reef Pipefishes:** *Corythoichthys haematopterus*, *Corythoichthys intestinalis*  
- **Freshwater Pipefishes:** *Doriyichthys boaja*, *Doriyichthys martensi*  
- **Cleaner Pipefishes:** *Doryhampus dactyliophorus*, *Doryrhamphus japonicus*  
- **Northern Pipefishes:** *Entelurus aequoreus*  
- **Mangrove Pipefishes:** *Hippichthys penicillus*  
- **River Pipefishes:** *Microphis brachyurus*  
- **Eel-like Pipefishes:** *Nerophis lumbriciformis*, *Nerophis ophidion*  
- **Alligator Pipefish:** *Syngnathoides biaculeatus*  
- **Common Pipefishes:** *Syngnathus abaster*, *Syngnathus acus*, *Syngnathus schlegeli*, *Syngnathus typhle*  
- **Deep-water Pipefishes:** *Trachyhampus serratus*  

---

2. Sex  
There are *2* sexes: *m* or *f*.  

---

3. Length  
- Mean Length: *175.57 mm*  
- Minimum Length: *51.5 mm*  
- Maximum Length: *415 mm*  

---

4. Brain Weight  
- Mean Brain Weight: *8.61 mg*  
- Minimum Brain Weight: *1.7 mg*  
- Maximum Brain Weight: *32.1 mg*  

---

5. Preservative  
There are *2* preservatives used:  
- Formalin  
- Ethanol  

---

6. Location  
There are **17** locations from which data was sampled. They are:  
*Japan, Malaysia, Micronesia, Marshall, Tokushima (Japan), Hahajima (Japan), Ioujima (Japan), Izu (Japan), Miyakejima (Japan), Sweden, Australia, Okinawa (Japan), Portugal, Otsuchi bay (Japan), Taiwan, Surugawan (Japan), Hyogo (Japan)*  


Since there are two species, we can further divide them into 2 excels to ease access.

In [29]:
# List of Seahorse species (Genus: Hippocampus)
seahorse_species = [
    "Hippocampus abdominalis", "Hippocampus comes", "Hippocampus hippocampus",
    "Hippocampus kuda", "Hippocampus spinosissimus", "Hippocampus trimaculatus"
]

# Separate the dataset into Seahorses and Pipefishes
brain_mass_seahorses_df = brain_mass_df[brain_mass_df["Species"].isin(seahorse_species)]
brain_mass_pipefishes_df = brain_mass_df[~brain_mass_df["Species"].isin(seahorse_species)]

# Save the separate Excel files
seahorses_file_path = "Data Pre-Processing/seahorses.xlsx"
pipefishes_file_path = "Data Pre-Processing/pipefishes.xlsx"

brain_mass_seahorses_df.to_excel(seahorses_file_path, index=False)
brain_mass_pipefishes_df.to_excel(pipefishes_file_path, index=False)

# Encoding Categorical Data for GA and RL

Now, as we want to implement both **Genetic Algorithm (GA)** and **Reinforcement Learning (RL)**, we need to encode the categorical data differently for each.

## Encoding for Genetic Algorithms (GA)  
**Use Label Encoding**  

GA works well with numerical representations rather than sparse matrices.  
One-hot encoding would create too many dimensions, making crossover and mutation inefficient.  
Label encoding assigns each category a unique integer, making crossover and mutation operations more meaningful.  

## Encoding for Reinforcement Learning (RL)  
**Use One-Hot Encoding**  

RL learns from state transitions, and categorical values shouldn't have a numerical hierarchy.  
One-hot encoding ensures the RL model treats each category as an independent entity.  
Label encoding may mislead RL into thinking higher numbers mean better categories (e.g., `Species_A = 1`, `Species_B = 2` suggests `B > A`).  


In [31]:
# Define file paths
seahorses_file_path = "Data Pre-Processing/seahorses.xlsx"
pipefishes_file_path = "Data Pre-Processing/pipefishes.xlsx"

# Load datasets
df_seahorses = pd.read_excel(seahorses_file_path)
df_pipefishes = pd.read_excel(pipefishes_file_path)

# For GA Algorithm, we will use label encoding.

In [ ]:
def apply_label_encoding(df):
    df_encoded = df.copy()
    label_encoders = {}
    for col in ["Species", "Preservative", "Location"]:
        le = LabelEncoder()
        df_encoded[col] = le.fit_transform(df_encoded[col])
        label_encoders[col] = le
    return df_encoded

df_seahorses_ga = apply_label_encoding(df_seahorses)
df_pipefishes_ga = apply_label_encoding(df_pipefishes)

# Save Label Encoded datasets for GA
seahorses_ga_file = "GA ALgorithm/seahorses_ga.xlsx"
pipefishes_ga_file = "GA ALgorithm/pipefishes_ga.xlsx"

df_seahorses_ga.to_excel(seahorses_ga_file, index=False)
df_pipefishes_ga.to_excel(pipefishes_ga_file, index=False)

In [51]:
df_seahorses_ga.head()

Species Sex  Length (mm)  Brain weight (mg)  Preservative  Location
0        0   m        135.0               12.6             1         0
1        0   f        184.5               15.8             1         0
2        0   f        146.7               10.1             1         0
3        0   f        156.6               20.2             1         0
4        0   m        151.0               10.9             1         0

In [50]:
df_pipefishes_ga.head()

Species Sex  Length (mm)  Brain weight (mg)  Preservative  Location
0        0   f        112.6               3.27             1         4
1        0   f        115.0               3.97             1         4
2        0   f        121.4               4.83             1         4
3        0   f        122.4               4.09             1         4
4        0   f        128.0               4.29             1         4

# For RL Algorithm, we will use label encoding.

In [ ]:
df_seahorses_rl = pd.get_dummies(df_seahorses, columns=["Species", "Preservative", "Location"], drop_first=True)
df_pipefishes_rl = pd.get_dummies(df_pipefishes, columns=["Species", "Preservative", "Location"], drop_first=True)

# Save One-Hot Encoded datasets for RL
seahorses_rl_file = "RL Algorithm/seahorses_rl.xlsx"
pipefishes_rl_file = "RL Algorithm/pipefishes_rl.xlsx"

df_seahorses_rl.to_excel(seahorses_rl_file, index=False)
df_pipefishes_rl.to_excel(pipefishes_rl_file, index=False)

In [52]:
df_seahorses_rl.head()

Sex  Length (mm)  Brain weight (mg)  Species_Hippocampus comes   
0   m        135.0               12.6                      False  \
1   f        184.5               15.8                      False   
2   f        146.7               10.1                      False   
3   f        156.6               20.2                      False   
4   m        151.0               10.9                      False   

   Species_Hippocampus hippocampus  Species_Hippocampus kuda   
0                            False                     False  \
1                            False                     False   
2                            False                     False   
3                            False                     False   
4                            False                     False   

   Species_Hippocampus spinosissimus  Species_Hippocampus trimaculatus   
0                              False                             False  \
1                              False                             False   
2                              False                             False   
3                              False                             False   
4                              False                             False   

   Preservative_Formalin  Location_Malaysia  
0                   True              False  
1                   True              False  
2                   True              False  
3                   True              False  
4                   True              False

In [53]:
df_pipefishes_rl.head()

Sex  Length (mm)  Brain weight (mg)  Species_Corythoichthys intestinalis   
0   f        112.6               3.27                                False  \
1   f        115.0               3.97                                False   
2   f        121.4               4.83                                False   
3   f        122.4               4.09                                False   
4   f        128.0               4.29                                False   

   Species_Doriyichthys boaja  Species_Doriyichthys martensi   
0                       False                          False  \
1                       False                          False   
2                       False                          False   
3                       False                          False   
4                       False                          False   

   Species_Doryhampus dactyliophorus  Species_Doryrhamphus japonicus   
0                              False                           False  \
1                              False                           False   
2                              False                           False   
3                              False                           False   
4                              False                           False   

   Species_Entelurus aequoreus  Species_Hippichthys penicillus  ...   
0                        False                           False  ...  \
1                        False                           False  ...   
2                        False                           False  ...   
3                        False                           False  ...   
4                        False                           False  ...   

   Location_Marshall  Location_Micronesia  Location_Miyakejima (Japan)   
0              False                False                        False  \
1              False                False                        False   
2              False                False                        False   
3              False                False                        False   
4              False                False                        False   

   Location_Okinawa (Japan)  Location_Otsuchi bay (Japan)  Location_Portugal   
0                     False                         False              False  \
1                     False                         False              False   
2                     False                         False              False   
3                     False                         False              False   
4                     False                         False              False   

   Location_Surugawan (Japan)  Location_Sweden  Location_Taiwan   
0                       False            False            False  \
1                       False            False            False   
2                       False            False            False   
3                       False            False            False   
4                       False            False            False   

   Location_Tokushima (Japan)  
0                       False  
1                       False  
2                       False  
3                       False  
4                       False  

[5 rows x 35 columns]

# Data Visualisation of head_shape.xlsx

In [35]:
# Load the head shape dataset
head_shape_path = "Data Pre-Processing/head_shape.xlsx"
head_shape_df = pd.read_excel(head_shape_path)

In [48]:
print(len(head_shape_df))

279


In [36]:
head_shape_df.head(5)
head_shape_df.head(-5)

0              1          2                3           4    
0                 1            23L         24                0       DIM=2  \
1    Chaematopterus  Cintestinalis     Dboaja  Ddactyliophorus  Djaponicus   
2        Sschlegeli        Styphle  Tserratus              NaN         NaN   
3         -0.356713       0.055463        NaN              NaN         NaN   
4         -0.359061      -0.013548        NaN              NaN         NaN   
..              ...            ...        ...              ...         ...   
269       -0.279988      -0.043781        NaN              NaN         NaN   
270       -0.283813       0.043850        NaN              NaN         NaN   
271       -0.101853       0.054450        NaN              NaN         NaN   
272       -0.062368       0.029681        NaN              NaN         NaN   
273        0.079042       0.103005        NaN              NaN         NaN   

            5           6             7       8             9      10   
0          NaN         NaN           NaN     NaN           NaN    NaN  \
1    Dmartensi  Eaequoreus  Habdominalis  Hcomes  Hhippocampus  Hkuda   
2          NaN         NaN           NaN     NaN           NaN    NaN   
3          NaN         NaN           NaN     NaN           NaN    NaN   
4          NaN         NaN           NaN     NaN           NaN    NaN   
..         ...         ...           ...     ...           ...    ...   
269        NaN         NaN           NaN     NaN           NaN    NaN   
270        NaN         NaN           NaN     NaN           NaN    NaN   
271        NaN         NaN           NaN     NaN           NaN    NaN   
272        NaN         NaN           NaN     NaN           NaN    NaN   
273        NaN         NaN           NaN     NaN           NaN    NaN   

              11              12             13           14               15   
0            NaN             NaN            NaN          NaN              NaN  \
1    Hpenicillus  Hspinosissimus  Htrimaculatus  Mbrachyurus  Nlumbriciformis   
2            NaN             NaN            NaN          NaN              NaN   
3            NaN             NaN            NaN          NaN              NaN   
4            NaN             NaN            NaN          NaN              NaN   
..           ...             ...            ...          ...              ...   
269          NaN             NaN            NaN          NaN              NaN   
270          NaN             NaN            NaN          NaN              NaN   
271          NaN             NaN            NaN          NaN              NaN   
272          NaN             NaN            NaN          NaN              NaN   
273          NaN             NaN            NaN          NaN              NaN   

            16        17     18            19  
0          NaN       NaN    NaN           NaN  
1    Nophidion  Sabaster  Sacus  Sbiaculeatus  
2          NaN       NaN    NaN           NaN  
3          NaN       NaN    NaN           NaN  
4          NaN       NaN    NaN           NaN  
..         ...       ...    ...           ...  
269        NaN       NaN    NaN           NaN  
270        NaN       NaN    NaN           NaN  
271        NaN       NaN    NaN           NaN  
272        NaN       NaN    NaN           NaN  
273        NaN       NaN    NaN           NaN  

[274 rows x 20 columns]

INFERENCE: UNECESSARY RIGHT NOW

# Data Visualisation of molecular_phylogeny.xlsx

In [37]:
# Load the molecular phylogeny dataset
molecular_phylogeny_path = "Data Pre-Processing/molecular_phylogeny.xlsx"
molecular_phylogeny_df = pd.read_excel(molecular_phylogeny_path)

In [46]:
print(len(molecular_phylogeny_df))

30


In [47]:
molecular_phylogeny_df.head(30)


Phylogeny
0                                              #NEXUS
1                                        BEGIN TREES;
2                  Title 'Trees from "lumbtree.nex"';
3                                   LINK Taxa = Taxa;
4                                           TRANSLATE
5                           [0] \t\t1 Chaematopterus,
6                            [1] \t\t2 Cintestinalis,
7                                   [2] \t\t3 Dboaja,
8                          [3] \t\t4 Ddactyliophorus,
9                               [4] \t\t5 Djaponicus,
10                               [5] \t\t6 Dmartensi,
11                              [6] \t\t7 Eaequoreus,
12                            [7] \t\t8 Habdominalis,
13                                  [8] \t\t9 Hcomes,
14                           [9] \t\t10 Hhippocampus,
15                                 [10] \t\t11 Hkuda,
16                           [11] \t\t12 Hpenicillus,
17                        [12] \t\t13 Hspinosissimus,
18                         [13] \t\t14 Htrimaculatus,
19                           [14] \t\t15 Mbrachyurus,
20                       [15] \t\t16 Nlumbriciformis,
21                             [16] \t\t17 Nophidion,
22                              [17] \t\t18 Sabaster,
23                                 [18] \t\t19 Sacus,
24                          [19] \t\t20 Sbiaculeatus,
25                            [20] \t\t21 Sschlegeli,
26                               [21] \t\t22 Styphle,
27                             [22] \t\t23 Tserratus;
28  TREE UNTITLED = (((1:0.272293,2:0.272293):0.59...
29                                               END;

INFERENCE: UNECESSARY RIGHT NOW